File: mobile/android/app/src/main/java/com/mobile/CustomWebViewManager.java
// import com.reactnativecommunity.webview.*;
// import com.facebook.react.ThemedReactContext;

// @ReactModule(name = CustomWebViewManager.REACT_CLASS)
// public class CustomWebViewManager extends RNCWebViewManager {
//   /* This name must match what we're referring to in JS */
//   protected static final String REACT_CLASS = "RCTCustomWebView";

//   protected static class CustomWebViewClient extends RNCWebViewClient { }

//   protected static class CustomWebView extends RNCWebView {
//     public CustomWebView(ThemedReactContext reactContext) {
//       super(reactContext);
//     }
//   }

//   @Override
//   protected RNCWebView createRNCWebViewInstance(ThemedReactContext reactContext) {
//     return new CustomWebView(reactContext);
//   }

//   @Override
//   public String getName() {
//     return REACT_CLASS;
//   }

//   @Override
//   protected void addEventEmitters(ThemedReactContext reactContext, WebView view) {
//     view.setWebViewClient(new CustomWebViewClient());
//   }
// }

File: mobile/android/app/src/main/java/com/mobile/CustomWebViewManager.kt

package com.mobile

import android.view.View
import com.facebook.react.ReactModule
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ReactShadowNode
import com.facebook.react.uimanager.ViewManager

@ReactModule(name = CustomWebViewManager.REACT_CLASS)
class CustomWebViewManager : RNCWebViewManager() {
    protected class CustomWebViewClient : RNCWebViewClient()
    protected inner class CustomWebView(reactContext: ThemedReactContext?) :
        RNCWebView(reactContext)

    override fun getExportedCustomDirectEventTypeConstants(): MutableMap<Any?, Any?>? {
        val superTypeConstants = super.getExportedCustomDirectEventTypeConstants()
        val export = superTypeConstants ?: MapBuilder.newHashMap<Any, Any?>()
        export["navigationCompleted"] = MapBuilder.of("registrationName", "onNavigationCompleted")
        return export
    }

    override fun createRNCWebViewInstance(reactContext: ThemedReactContext?): RNCWebView {
        return CustomWebView(reactContext)
    }

    override fun addEventEmitters(reactContext: ThemedReactContext, view: WebView) {
        view.webViewClient = CustomWebViewClient()
    }

    companion object {
        /* This name must match what we're referring to in JS */
        const val REACT_CLASS = "RCTCustomWebView"
    }

    @ReactProp(name = "finalUrl")
    fun setFinalUrl(view: WebView, url: String?) {
        (view as CustomWebView).finalUrl = url
    }
}

protected class CustomWebViewClient : RNCWebViewClient() {
    override fun shouldOverrideUrlLoading(view: WebView, url: String?): Boolean {
        val shouldOverride: Boolean = super.shouldOverrideUrlLoading(view, url)
        val finalUrl: String? = (view as CustomWebView).finalUrl
        if (!shouldOverride && url != null && finalUrl != null && url == finalUrl) {
            val params: WritableMap = Arguments.createMap()
            dispatchEvent(view, NavigationCompletedEvent(view.getId(), params))
        }
        return shouldOverride
    }
}
